### PR TITLE
Generalized SegmentFromPoint to include camera info and average depth

### PR DESCRIPTION
**Pull request by amalnanavati**
# Description

This PR generalizes the SegmentFromPoint node to also send the camera info and the average depth of each mask in the result. This is so that AcquireAction has all the data it needs without needing to subscribe to camera topics.

# Testing procedure

On Lovelace, I ran the web app (PR[#79](https://github.com/personalrobotics/feeding_web_interface/pull/79)) and SegmentFromPoint (through the launchfile). I modified the web app to log the response from the SegmentFromPoint action (on [line 162](https://github.com/personalrobotics/feeding_web_interface/blob/a44bc0d23523e64987972b1fb90edd7b9afb6b14/feedingwebapp/src/Pages/Home/MealStates/BiteSelection.jsx#L162)) to log the response. I then used the web app to click a food item. I verified (a) that the depth data made sense (it was 0.426m for the masks that were close to the food item, which was correct), and (b) that the camera info is correct for the camera color optical frame (compared it to an echo of that topic).

# Before opening a pull request
- [x] Format your code using [black formatter](https://black.readthedocs.io/en/stable/) `python3 -m black .`
- [x] Run your code through [pylint](https://pylint.readthedocs.io/en/latest/) and address (most) warnings/errors: `pylint --recursive=y .`

# Before Merging
- [ ] `Squash & Merge`

### CODE REVIEW

File: ada_feeding_msgs/msg/Mask.msg [hunk 1]
```diff
@@ -6,8 +6,13 @@ sensor_msgs/RegionOfInterest roi
...
+
+# The average of the depth camera readings over the mask (in m)
+float64 average_depth
+
 # An arbitrary ID that defines the segmented item
 string item_id
+
 # A score that indicates how confident the segemntation algorithm is in 
 # this mask.
 float64 confidence
```
Comment (egordon) [inline ada_feeding_msgs/msg/Mask.msg:18]: Nit: best to add a newline at EOF

File: ada_feeding_perception/ada_feeding_perception/segment_from_point.py [hunk 2]
```diff
@@ -78,7 +78,28 @@ def __init__(self) -> None:
...
+            1,
+        )
+        self.camera_info = None
+        self.camera_info_lock = threading.Lock()
+
+        # Subscribe to the aligned depth image topic, to store the latest depth image
+        # NOTE: We assume this is in the same frame as the RGB image
+        self.depth_image_subscriber = self.create_subscription(
+            Image,
+            "/camera/aligned_depth_to_color/image_raw",
```
Comment (egordon) [inline ada_feeding_perception/ada_feeding_perception/segment_from_point.py:95]: Are all these topic names hard-coded? 

A better tactic would be to have these strings be "image", "aligned_depth", and "camera_info", and remap them in a roslaunch.

This isn't required right now, but add it as an Issue before Merging so we don't forget.

File: ada_feeding_perception/ada_feeding_perception/segment_from_point.py [hunk 3]
```diff
@@ -191,6 +212,28 @@ def initialize_food_segmentation(self) -> None:
...
+        Store the latest depth image message.
+
+        Parameters
+        ----------
+        msg: The depth image message.
+        """
+        with self.latest_depth_img_msg_lock:
+            self.latest_depth_img_msg = msg
+
     def image_callback(self, msg: Image) -> None:
```
Comment (egordon) [inline ada_feeding_perception/ada_feeding_perception/segment_from_point.py:237]: As above, not needed for this PR, but flagging as an Issue. We should make this `msg: Union[CompressedImage, Image]` to support either arg type.

That can apply to `segment_image` as well, using `image_msg.__class__ == CompressedImage` as an if statement to swap between the two bridge functions.

And for the subscriber, you can glean which topic type (and also check the topic type at runtime) by calling `self.get_topic_names_and_types()`, [searching through the list of tuples for the topic name](https://stackoverflow.com/questions/2917372/how-to-search-a-list-of-tuples-in-python), and checking the type name for `sensor_msgs/CompressedImage` or `sensor_msgs/Image`. Perhaps breaking it off into a helper function.

File: ada_feeding_perception/ada_feeding_perception/segment_from_point.py [hunk 4]
```diff
@@ -260,11 +307,22 @@ def segment_image(
...
+
+        # Get the latest depth image
+        with self.latest_depth_img_msg_lock:
+            depth_img_msg = self.latest_depth_img_msg
 
         # Convert the image to OpenCV format
         image = self.bridge.imgmsg_to_cv2(image_msg, desired_encoding="bgr8")
         image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
 
+        # Convert the depth image to OpenCV format
```
Comment (egordon) [inline ada_feeding_perception/ada_feeding_perception/segment_from_point.py:321]: Nit: Probably best to reiterate as a comment here that the data format is 16bit Integer mm.

File: ada_feeding_perception/ada_feeding_perception/segment_from_point.py (hunk 2 above)
Comment (egordon) [reply ada_feeding_perception/ada_feeding_perception/segment_from_point.py:95]: (Unless it's fast to update)

File: ada_feeding_perception/ada_feeding_perception/segment_from_point.py [hunk 5]
```diff
@@ -287,6 +345,10 @@ def segment_image(
             # the seed point
             # TODO: Thorughly test this, in case we need to add more mask cleaning!
             cleaned_mask = get_connected_component(mask, seed_point)
```
Comment (egordon) [inline ada_feeding_perception/ada_feeding_perception/segment_from_point.py:347]: Just recording a thought: consider erosion/dilation to remove small imperfections in the mask.

**Review by egordon — APPROVED** [review]
A few nits and a few things to add as Issues for later (sorry, it's a product of me not getting a chance to review earlier PRs), but I think aren't detrimental to go in now.

File: ada_feeding_perception/ada_feeding_perception/segment_from_point.py (hunk 5 above)
Comment (amalnanavati) [reply ada_feeding_perception/ada_feeding_perception/segment_from_point.py:347]: Added it as an in-text comment.

File: ada_feeding_perception/ada_feeding_perception/segment_from_point.py (hunk 3 above)
Comment (amalnanavati) [reply ada_feeding_perception/ada_feeding_perception/segment_from_point.py:237]: Added it in, yet to test.

**Comment by amalnanavati** [comment]
Addressed all comments. I'll test it on LoveLace and, if that works, will merge.

**Comment by amalnanavati** [comment]
Everything works on LoveLace, including both CompressedImage and Image for the color image stream.

**Pull request merged**